### DISCUSSION TITLE
[@types/aws-lambda] Migrating to CDK version 2 results in test cases being failed

### DISCUSSION THREAD

**Discussion by jatinmehrotra** [DefinitelyTyped/DefinitelyTyped#61401] [Issues with a @types package]
I converted my cdk infrastructure code from version 1 to version 2. It was a successful conversion.

However, some of the test cases are now being failed. I haven't touched any other code except changing references from version to version 2.

Possible causes :- The infrastructure uses [apigatewayv2](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-apigatewayv2-alpha-readme.html), [apigatewayv2-integrations](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-apigatewayv2-integrations-alpha-readme.html), [apigatewayv2-authorizers](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-apigatewayv2-authorizers-alpha-readme.html) ( alpha version of these packages)

Since I have not written the code, i cannot help you with the development code. However, i can provide the code where test cases are being failed.

Package for lambda is being used as @types/aws-lambda https://www.npmjs.com/package/@types/aws-lambda

```
import { BookDetailUseCaseInput } from '@use-cases/suken-account/detail-book'
import { IsString, Length, Matches, MinLength } from 'class-validator'
import { APIGatewayProxyEvent } from 'aws-lambda'

export class BookDetailRequest implements BookDetailUseCaseInput {
  @IsString({
    message: 'something',
  })
  @Length(1, 30, {
    message: ''something。',
  })
  @Matches(/^[a-zA-Z_]\w*$/, {
    message: ''something。',
  })
  readonly bookId: string

  @IsString({
    message: ''something。',
  })
  @MinLength(1, {
    message: ''something。',
  })
  readonly token: string

  @IsString({
    message: 'something。',
  })
  @MinLength(1, {
    message: ''something。',
  })
  readonly appVersion: string

  constructor(event: APIGatewayProxyEvent) {
    this.bookId = event.pathParameters!.book_id
    this.token = event.headers.authorization?.split('Bearer')[1]?.trimLeft()
    this.appVersion = event.headers['app-viewer-version']
  }
}
```
error message during test run

  35     this.token = event.headers.authorization?.split('Bearer')[1]?.trimLeft()
           ~~~~~~~~~~
    src/modules/validators/wdwdwd-account/wdwdwdwd-request.ts:36:5 - error TS2322: Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.

    36     this.appVersion = event.headers['app-viewer-version']
I think APIGatewayProxyEvent is sending values in different format in cdk version 2

Is there something needs to be changed for lambda when migration to cdk version 2?

[a comment by an automated account is not shown]

**Comment by simonbuchan** [comment]
This code is obviously wrong by itself: token is `string` and thus cannot be `undefined`, and the assignment is from an optional chaining (`?.`) expression, making it clear it may be `undefined`.

Raise this issue with wherever you got the code, or simply replace those `?` with `!`, like with the first property, or change the types to `string | undefined`, depending on whether the values might actually not be present at runtime.
